Source/DafnyCore/Resolver/PreType/UnderspecificationDetector.cs

```diff
@@ -0,0 +1,584 @@
+                    rolemodel.IsGhost ? "ghost" : "non-ghost", dtor.EnclosingCtors[0].Name,
+                    other.IsGhost ? "ghost" : "non-ghost", dtor.EnclosingCtors[i].Name);
+                }
+              }
+            }
+          }
+          foreach (var ctor in dd.Ctors) {
+            CheckParameterDefaultValues(ctor.Formals, dd);
+          }
+          CheckMembers(dd);
+
+        } else if (d is TopLevelDeclWithMembers md) {
+          CheckMembers(md);
+        }
+      }
+    }
+
+    private void CheckMembers(TopLevelDeclWithMembers cl) {
+      Contract.Requires(cl != null);
+      foreach (var member in cl.Members) {
+        var prevErrCnt = ErrorCount;
+        CheckMember(member);
+        if (prevErrCnt == ErrorCount) {
+          if (member is Method method) {
+            if (method.Body != null) {
+              CheckExpressions(method.Body, method);
+            }
+          } else if (member is Function function) {
+            if (function.ByMethodBody != null) {
+              var m = function.ByMethodDecl;
+              CheckExpressions(m.Body, m);
+            }
+          }
+          if (prevErrCnt == ErrorCount && member is ICodeContext) {
+            member.SubExpressions.Iter(e => CheckExpression(e, (ICodeContext)member));
+          }
+        }
+      }
+    }
+
+    private void CheckMember(MemberDecl member) {
+      if (member is ConstantField field) {
+        if (field.Rhs != null) {
+          CheckExpression(field.Rhs, new NoContext(member.EnclosingClass.EnclosingModuleDefinition));
+        }
+        CheckPreType(field.PreType, new NoContext(member.EnclosingClass.EnclosingModuleDefinition), field.tok, "const");
+
+      } else if (member is Method method) {
+        CheckParameterDefaultValues(method.Ins, method);
+        method.Req.Iter(mfe => CheckAttributedExpression(mfe, method));
+        CheckSpecFrameExpression(method.Mod, method);
+        method.Ens.Iter(mfe => CheckAttributedExpression(mfe, method));
+        CheckSpecExpression(method.Decreases, method);
+        if (method.Body != null) {
+          CheckExpressions(method.Body, method);
+        }
+
+      } else if (member is Function) {
+        var f = (Function)member;
+        CheckParameterDefaultValues(f.Formals, f);
+        var errorCount = ErrorCount;
+        f.Req.Iter(e => CheckExpression(e.E, f));
+        f.Ens.Iter(e => CheckExpression(e.E, f));
+        f.Reads.Iter(fe => CheckExpression(fe.E, f));
+        CheckSpecExpression(f.Decreases, f);
+        if (f.Body != null) {
+          CheckExpression(f.Body, f);
+        }
+        if (errorCount == ErrorCount) {
+          if (f is ExtremePredicate cop) {
+            CheckMember(cop.PrefixPredicate);
+          } else if (f.ByMethodDecl != null) {
+            CheckMember(f.ByMethodDecl);
+          }
+        }
+      }
+    }
+
+    private void CheckPreType(PreType preType, ICodeContext codeContext, IToken tok, string what) {
```

Comment:
`what` could perhaps be renamed to e.g. `stringType`

Reply:
I renamed it to `whatIsBeingChecked`